README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# shinyTime <a href='https://burgerga.shinyapps.io/shinyTimeExample/' target="_blank"><img src='man/figures/timeInput.png' align="right" height="138.5" /></a>

<!-- badges: start -->

[![R-CMD-check](https://github.com/burgerga/shinyTime/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/burgerga/shinyTime/actions/workflows/R-CMD-check.yaml)
[![](http://www.r-pkg.org/badges/version/shinyTime)](https://cran.r-project.org/package=shinyTime)
[![](http://cranlogs.r-pkg.org/badges/last-month/shinyTime)](https://cran.r-project.org/package=shinyTime)
<!-- badges: end -->

## Overview

shinyTime provides a `timeInput` widget for Shiny. This widget allows
intuitive time input in the `hh:mm:ss` or `hh:mm` (24-hour) format by
using a separate numeric input for each time component. Setting and
getting of the time in R is done with date-time objects.

## Installation

``` r
# Install from CRAN
install.packages("shinyTime")
```

## Usage

As the `shinyTime` package mimics the existing shiny functionality,
using the package is easy:

``` r
ui <- fluidPage(
  # Using the default time 00:00:00
  timeInput("time1", "Time:"),

  # Set to current time
  timeInput("time2", "Time:", value = Sys.time()),

  # Set to custom time 
  timeInput("time3", "Time:", value = strptime("12:34:56", "%T")),
 
  # Set to custom time using hms
  timeInput("time4", "Time:", value = hms::as_hms("23:45:07")),

  # Set to custom time using character string
  timeInput("time5", "Time:", value = "21:32:43"),

  # Use hh:mm format
  timeInput("time6", "Time:", seconds = FALSE),

  # Use multiples of 5 minutes
  timeInput("time7", "Time:", minute.steps = 5)
)
```

Note that setting an initial value can be done with
[`date-time`](https://www.rdocumentation.org/packages/base/topics/DateTimeClasses)
(in the same way as setting a date in `dateInput` can be done with a
`Date` object), but also with an
[`hms::hms`](https://hms.tidyverse.org/reference/hms.html) object or
character string in `hh:mm:ss` format.

The value retrieved will be a `date-time` object (`POSIXlt`). You need
to convert it to character to be able to show the time, as the default
character representation does not include time. For example:

``` r
server <- function(input, output) {
  # Print the time in hh:mm:ss everytime it changes
  observe(print(strftime(input$time1, "%T")))
  
  # Print the time in hh:mm everytime it changes
  observe(print(strftime(input$time4, "%R")))
}
```

For a demo, visit the [online example
app](https://burgerga.shinyapps.io/shinyTimeExample/) or try the
`shinyTime::shinyTimeExample()` function.


NEWS.md
# shinyTime (development version)

# shinyTime 1.0.3

## Minor changes

- Add support for `hms::hms` and character input values (#18)

# shinyTime 1.0.2

## Minor changes

- Add information on `timeInput` return type (#14)

## Bug fixes

- Replaced outdated shiny utils label functions (#16)

- Added correct timezone handling to `roundTime` function (#11)

# shinyTime 1.0.1

## Bug fixes

- Removed use of ES6 that caused `shinyTime` to fail on older browsers (#8)

# shinyTime 1.0.0

## Major changes 

- Input time with minutes rounded to the nearest user-specified multiple

# shinyTime 0.2.1

## Minor changes

- Updated style to match other shiny inputs

- Added input validation so that the time returned is always a valid 
time (invalid values are set to 0)

# shinyTime 0.2.0

## Minor changes

- Added support for input in the %H:%M format (without seconds)

## Bug fixes

- Fixed padding problem on keyboard input

- Keyboard input now properly causes change events

# shinyTime 0.1.0

- Initial release


tests/testthat.R
library(testthat)
library(shinyTime)

test_check("shinyTime")


tests/testthat/test-round-time.R
test_that("rounding time works", {
  inputs <- c("00:00:00", "00:02:00", "23:59:00", "23:56:00")
  outputs <- c("00:00:00", "00:00:00", "00:00:00", "23:55:00")
  expect_equal(strftime(roundTime(timeStringToDate(inputs), 5), format = "%T"), outputs)

})

test_that("non-integer minutes fail", {
  expect_error(roundTime(Sys.time, 0.5))
})

test_that("hms works", {
  expect_equal(strftime(roundTime(hms::as_hms("09:00:01"), minutes = 1), format = "%T"), "09:00:00")
})


R/zzz.R
#' Sets up the package when it's loaded
#'
#' Adds the content of www to shinyTime/, and registers an inputHandler to massage the output from
#' JavaScript into an R structure.
#'
#' @importFrom shiny addResourcePath registerInputHandler
#'
#' @noRd
#'
.onLoad <- function(libname, pkgname) {
  # Add directory for static resources
  addResourcePath('shinyTime', system.file('www', package='shinyTime', mustWork = TRUE))
  # Do some processing on the data we get from javascript before we pass it on to R
  registerInputHandler('my.shiny.timeInput', function(data, ...) {
    # Replace NULL by 0
    data[sapply(data, is.null)] <- 0
    # Convert to time object
    timeListToDate(data)
  })
}

#' Cleans up when package is unloaded
#'
#' Reverses the effects from .onLoad
#'
#' @importFrom shiny removeInputHandler
#'
#' @noRd
#'
.onUnload <- function(libpath) {
  removeInputHandler('my.shiny.timeInput')
}

R/utils.R
# Some utility functions

# Copied from shiny/R/input-utils.R
shinyInputLabel <- function(inputId, label = NULL, control = FALSE) {
  classes <- c(
    if (is.null(label)) "shiny-label-null",
    if (control) "control-label"
  )
  tags$label(
    label,
    `for` = inputId,
    class = if (!is.null(classes)) paste(classes, collapse = " ")
  )
}

# Given a vector or list, drop all the NULL items in it
# Copied from shiny/R/utils.R
dropNulls <- function(x) {
  x[!vapply(x, is.null, FUN.VALUE=logical(1))]
}

dateToTimeList <- function(value){
  if(is.null(value)) return(NULL)
  posixlt_value <- unclass(as.POSIXlt(value))
  time_list <- lapply(posixlt_value[c('hour', 'min', 'sec')], function(x) {
    sprintf("%02d", trunc(x))
  })
  return(time_list)
}

timeListToDate <- function(value) {
  timeStringToDate(paste(c(value$hour, value$min, value$sec), collapse = ':'))
}

timeStringToDate <- function(string) {
  strptime(string, format = "%T")
}

getDefaultTime <- function(){
  timeStringToDate("00:00:00")
}

# From ?is.integer
is.wholenumber <- function(x, tol = .Machine$double.eps^0.5)  abs(x - round(x)) < tol

roundTime <- function(time, minutes) {
  stopifnot(any(class(time) %in% c("POSIXt", "hms")))
  stopifnot(is.wholenumber(minutes))
  if("hms" %in% class(time)) {
    # if hms reset to local time zone instead of UTC
    # works by getting the hour, minute, sec components out, and constructing a new POSIXlt object
    # a bit inefficient, but only happens once for each timeInput
    time <- timeListToDate(dateToTimeList(time))
  }
  time <- as.POSIXct(time)
  # Copied from plyr:::round_any.numeric
  round_any <- function(x, accuracy, f=round){f(x/accuracy) * accuracy}
  s <- round_any(unclass(time), 60 * minutes)
  # Inspired by lubridate::origin
  structure(s, class = c("POSIXct", "POSIXt"), tzone = attr(time, "tzone"))
}


inst/example/app.R
#
# This is a Shiny web application.
#
# Find out more about building applications with Shiny here:
#
#    http://shiny.rstudio.com/
#

library(shiny)
library(shinyTime)

ui <- fluidPage(

   titlePanel("shinyTime Example App"),

   sidebarLayout(
      sidebarPanel(
        timeInput("time_input1", "Enter time", value = strptime("12:34:56", "%T")),
        timeInput("time_input2", "Enter time (5 minute steps)", value = strptime("12:34:56", "%T"), minute.steps = 5),
        actionButton("to_current_time", "Current time")
      ),

      mainPanel(
        textOutput("time_output1"),
        textOutput("time_output2")
      )
   )
)

server <- function(input, output, session) {
  output$time_output1 <- renderText(strftime(input$time_input1, "%T"))
  output$time_output2 <- renderText(strftime(input$time_input2, "%R"))

  observeEvent(input$to_current_time, {
    updateTimeInput(session, "time_input1", value = Sys.time())
    updateTimeInput(session, "time_input2", value = Sys.time())
  })

}

shinyApp(ui, server)



inst/www/input_binding_time.js
// Enclose in IEFE
(function() {
// Escape jQuery selector metacharacters: !"#$%&'()*+,./:;<=>?@[\]^`{|}~
// Copied from shiny/srcjs/utils.js
var $escape = function(val) {
  return val.replace(/([!"#$%&'()*+,.\/:;<=>?@\[\\\]^`{|}~])/g, '\\$1');
};

var findLabelForElement = function(el) {
  return($(el).parent().find('label[for="' + $escape(el.id) + '"]'));
};

var clamp = function(num, min, max) {
  return Math.min(Math.max(num, min), max);
};

var inRange = function(num, min, max) {
  return (num >= min && num <= max);
};

var zeroPad = function(num) {
  return (num < 10 ? '0' + num: num);
};

var correctInputValue = function(el) {
  var $el = $(el);
  var val = parseFloat($el.val());
  // Check if number is integer, if so put to fixed form (0.1e1 will become 1), else make 0
  var newVal = (val % 1 === 0) ? val.toFixed() : 0;
  // Make 0 if out of range, alternative would be clamping.
  if($el.hasClass('shinytime-hours')) {
   newVal = inRange(newVal, 0, 23) ? newVal : 0;
  } else {
   newVal = inRange(newVal, 0, 59) ? newVal : 0;
  }
  // correct minute step
  if($el.hasClass('shinytime-mins')){
    var step = $el.attr("step");
    if(step && step != 1) {
      newVal = inRange(newVal, 0, 55) ? Math.round(newVal / step) * step : 55;
    }
  }
  // Zero pad and update value
  $el.val(zeroPad(newVal));
};

// From https://stackoverflow.com/a/48512262/1439843
var roundTime = function(value, minutesToRound) {

  var roundTimeDate = function(date) {
    var ms = 1000 * 60 * minutesToRound; // convert minutes to ms
    return new Date(Math.round((new Date(date)).getTime() / ms) * ms);
  };

  roundedDate = roundTimeDate((new Date()).setHours(value.hour, value.min, value.sec));
  return {hour: zeroPad(roundedDate.getHours()), min: zeroPad(roundedDate.getMinutes()), sec: zeroPad(roundedDate.getSeconds())};
};

var timeInputBinding = new Shiny.InputBinding();

$.extend(timeInputBinding, {
  find: function(scope) {
    return $(scope).find('.my-shiny-time-input');
  },
  getValue: function(el) {
    var $inputs = $(el).find('input');
    var values = [];
    $inputs.each(function() {
      numberVal = $(this).val();
      // Number parsing logic taken from shiny/srcjs/input_binding_number.js
      if (/^\s*$/.test(numberVal))  // Return null if all whitespace
        values.push(null);
      else if (!isNaN(numberVal))   // If valid Javascript number string, coerce to number
        values.push(+numberVal);
      else
        values.push(numberVal);
    });
    return {
      hour: values[0],
      min:  values[1],
      sec:  (values.length > 2) ? values[2] : 0
    };
  },
  setValue: function(el, value) {
    var $inputs = $(el).find('input');
    minuteSteps = $inputs.eq(1).attr("step");
    if(minuteSteps && minuteSteps != 1) value = roundTime(value, minuteSteps);
    $inputs.eq(0).val(value.hour);
    $inputs.eq(1).val(value.min);
    $inputs.eq(2).val(value.sec);
  },
  receiveMessage: function(el, data) {
    // To get updateTimeInput working
    if (data.hasOwnProperty('label')) findLabelForElement(el).text(data.label);
    if (data.hasOwnProperty('value')) this.setValue(el, data.value);
    $(el).trigger('change');
  },
  getState: function(el) {
    return {
      label: findLabelForElement(el).text(),
      value: this.getValue(el)
    };
  },
  getType: function() {
    // Necessary to get the registerInputHandler working
    return "my.shiny.timeInput";
  },
  subscribe: function(el, callback) {
    $(el).on("change.timeInputBinding", function(e) {
      correctInputValue(e.target);
      callback();
    });
  },
  unsubscribe: function(el) {
    $(el).off(".timeInputBinding");
  }
});

Shiny.inputBindings.register(timeInputBinding, 'my.shiny.timeInput');
})();


R/input-time.R
#' @keywords internal
"_PACKAGE"

#' Create a time input
#'
#' Creates a time widget that consists of separate numeric inputs for the hours, minutes, and
#' seconds. The input and output values of the time widget are instances of
#' \code{\link{DateTimeClasses}}, these can be converted to and from character strings with
#' \code{\link{strptime}} and \code{\link{strftime}}.
#' Additionally, the input can be specified as a character string in the `hh:mm:ss` format or an
#' \code{\link[hms]{hms}} class. For a simple example app see \code{\link{shinyTimeExample}}.
#'
#' @inheritParams shiny::textInput
#' @param value The desired time value. Must be a instance of \code{\link{DateTimeClasses}}.
#' @param seconds Show input for seconds. Defaults to TRUE.
#' @param minute.steps Round time to multiples of \code{minute.steps} (should be a whole number).
#' If not NULL sets \code{seconds} to \code{FALSE}.
#'
#' @returns Returns a \code{POSIXlt} object, which can be converted to
#' a \code{POSIXct} object with \code{as.POSIXct} for more efficient storage.
#'
#' @family shinyTime functions
#' @seealso \code{\link{strptime}}, \code{\link{strftime}}, \code{\link{DateTimeClasses}}
#'
#' @examples
#' ## Only run examples in interactive R sessions
#' if (interactive()) {
#'
#' ui <- fluidPage(
#'   # Default value is 00:00:00
#'   timeInput("time1", "Time:"),
#'
#'   # Set to current time
#'   timeInput("time2", "Time:", value = Sys.time()),
#'
#'   # Set to custom time
#'   timeInput("time3", "Time:", value = strptime("12:34:56", "%T")),
#'
#'   # Set to custom time using hms
#'   timeInput("time4", "Time:", value = hms::as_hms("23:45:07")),
#'
#'   # Set to custom time using character string
#'   timeInput("time5", "Time:", value = "21:32:43"),
#'
#'   # Use hh:mm format
#'   timeInput("time6", "Time:", seconds = FALSE),
#'
#'   # Use multiples of 5 minutes
#'   timeInput("time7", "Time:", minute.steps = 5)
#' )
#'
#' shinyApp(ui, server = function(input, output) { })
#' }
#'
#' @importFrom htmltools tagList singleton tags
#' @export
timeInput <- function(inputId, label, value = NULL, seconds = TRUE, minute.steps = NULL) {
  if(is.null(value)) value <- getDefaultTime()
  if(is.character(value)) value <- strptime(value, format = "%T")
  if(!is.null(minute.steps)) {
    stopifnot(is.wholenumber(minute.steps))
    seconds = F
    value <- roundTime(value, minute.steps)
  }
  value_list <- dateToTimeList(value)
  style <- "width: 8ch"
  input.class <- "form-control"
  tagList(
    singleton(tags$head(
      tags$script(src = "shinyTime/input_binding_time.js")
    )),
    tags$div(id = inputId, class = "my-shiny-time-input form-group shiny input-container",
      shinyInputLabel(inputId, label, control = TRUE),
      tags$div(class = "input-group",
        tags$input(type="number", min="0", max="23", step="1", value = value_list$hour,
                   style = style, class = paste(c(input.class, 'shinytime-hours'), collapse = " ")),
        tags$input(type="number", min="0", max="59", step=minute.steps, value = value_list$min,
                   style = style, class = paste(c(input.class, 'shinytime-mins'), collapse = " ")),
        if(seconds) tags$input(type="number", min="0", max="59", step="1", value = value_list$sec,
                   style = style, class = paste(c(input.class, 'shinytime-secs'), collapse =  " ")) else NULL
      )
    )
  )
}

#' Change a time input on the client
#'
#' Change the label and/or value of a time input
#'
#' @inheritParams shiny::updateTextInput
#' @param value The desired time value. Must be a instance of \code{\link{DateTimeClasses}}.
#'
#' @family shinyTime functions
#'
#' @examples
#' ## Only run examples in interactive R sessions
#' if (interactive()) {
#'
#' ui <- fluidPage(
#'   timeInput("time", "Time:"),
#'   actionButton("to_current_time", "Current time")
#' )
#'
#' server <- function(input, output, session) {
#'   observeEvent(input$to_current_time, {
#'     updateTimeInput(session, "time", value = Sys.time())
#'   })
#' }
#'
#' shinyApp(ui, server)
#' }
#'
#' @export
updateTimeInput <- function(session, inputId, label = NULL, value = NULL) {
  value <- dateToTimeList(value)
  message <- dropNulls(list(label=label, value = value))
  session$sendInputMessage(inputId, message)
}

#' Show the shinyTime example app
#'
#' Run a simple shiny app demonstrating the shinyTime functionality.
#'
#' @family shinyTime functions
#' @importFrom shiny runApp
#' @export
shinyTimeExample <- function() {
  runApp(system.file('example', package='shinyTime', mustWork=T), display.mode='showcase')
}
